Do it yourself:
design a case matching with if else code, for the following students marks gradation:
1. below 30% - Grade F
2. Between 30 to 40% - Grade E
3. Between 41 to 50% - Grade D
4. Between 51-60% - Grade C
5. Between 61-70% - Grade B
6. Between 71-80% - Grade AB
7. Between 81-90% - Grade A
8. Between 91-100% - Grade AA
print the grade when a student enters marks.

In [ ]:
marks = float(input("Enter student marks (0-100): "))

if marks < 0 or marks > 100:
    print("Invalid marks")

elif marks < 30:
    print("Grade F")

elif marks <= 40:
    print("Grade E")

elif marks <= 50:
    print("Grade D")

elif marks <= 60:
    print("Grade C")

elif marks <= 70:
    print("Grade B")

elif marks <= 80:
    print("Grade AB")

elif marks <= 90:
    print("Grade A")

else:
    print("Grade AA")


Enter student marks (0-100): 56
Grade C


Python Loops
Python has two primitive loop commands:

while loops,
for loops

The while Loop: With the while loop we can execute a set of statements as long as a condition is true.

In [3]:
i = 1 #initialization of the variable
while i <= 7:  #syntax
#loop condition
  print(i)  #if we do not give increment statement the loop is infinite
  i = i+1
print("the code is executed.")

1
2
3
4
5
6
7
the code is executed.


In [ ]:
battery = int(input("Enter the battery % "))

while battery < 100:
    print("Charging... Battery at", battery, "%")
    battery = battery + 10

print("Battery fully charged!")

Enter the battery % 24
Charging... Battery at 24 %
Charging... Battery at 34 %
Charging... Battery at 44 %
Charging... Battery at 54 %
Charging... Battery at 64 %
Charging... Battery at 74 %
Charging... Battery at 84 %
Charging... Battery at 94 %
Battery fully charged!


If and while together

In [ ]:
# Take a number as input
num = int(input("Enter a number less than 10: "))

# Check if the number is less than 10
if num < 10:
    count = 1
    while count <= 10:
        print("I learn while")
        count += 1
else:
    print("Number is not less than 10")


Enter a number less than 10: 9
I learn while
I learn while
I learn while
I learn while
I learn while
I learn while
I learn while
I learn while
I learn while
I learn while


Use of while

In [ ]:
# Take a number as input
num = int(input("Enter a number less than 10: "))

count = 1

while count <= num:
    print("I learn while")
    count += 1


Enter a number less than 10: 5
I learn while
I learn while
I learn while
I learn while
I learn while


The break Statement
With the break statement we can stop the loop even if the while condition is true:

In [ ]:
i = 1
while i < 6:
  print(i)
  if i == 3:
    break
  i += 1

1
2
3


The continue Statement
With the continue statement we can stop the current iteration, and continue with the next:

In [ ]:
i = 0
while i < 6:
  i += 1
  if i == 3:
    continue
  print(i)


1
2
4
5
6


The else Statement
With the else statement we can run a block of code once when the condition no longer is true:

In [ ]:
i = 1
while i < 6:
  print(i)
  i += 1
else:
  print("i is no longer less than 6")

1
2
3
4
5
i is no longer less than 6


Write a program to print numbers from 1 to 10 using a while loop.

Print all even numbers from 1 to 20 using a while loop.

In [ ]:
i = 1
while i <=20:
  if i%2==0:
    print("This is an even number", i)


Take a number as input and print "I love Python" that many times using a while loop.

Write a program to print numbers from 10 to 1 using a while loop.

Write a program to find the sum of numbers from 1 to n, where n is given by the user. for each of the numbers, sum should be written.

for 0, it is 0,

for 1 it is (0+1)

for 2 it is (0+1+2)

for 3 it is (0+1+2+3) and so on

What if I want to print only the final sum.

Take a number as input and print its multiplication table using a while loop.

Take a number as input and count how many digits it has using a while loop.

In [ ]:
num = int(input("Enter your number:"))
count = 0
if num == 0:
    count = 1
else:
  while num>0:
    count = count+1
    num = num//10
    #This line performs integer (floor) division of num by 10 and stores the result back in num.
#What // means , // → floor division, It divides and removes the decimal part
#The result is always an integer (rounded down)
#example 123/10 = 12
print("Digit Count is:", count)


Enter your number:4556
Digit Count is: 4


Take a number as input and reverse it using a while loop.

In [ ]:
num = int(input("Enter a number: "))

rev = 0

while num > 0:
    digit = num % 10        # Get last digit
    rev = rev * 10 + digit # Add digit to reversed number
    num //= 10             # Remove last digit

print("Reversed number:", rev)


Enter a number: 4563
Reversed number: 3654


Keep asking the user to enter a number until they enter 7.
Print "Correct guess!" when they do.

Keep asking for a password using a while loop until the correct password is entered.

Take a number as input and find its factorial using a while loop.

In [ ]:
n = int(input("Enter a number:"))
fact = 1
if(n==0):
  print("factorial is ", fact)
elif(n==1):
  print("factorial is .", fact)
else:
  while(n>=1):
    fact = fact*n
    n = n-1
print(f"The factorial is {fact}")


Enter a number:5
The factorial is 120


Take a number and check if it is a palindrome using a while loop.

In [ ]:
num = int(input("Enter a number: "))

original = num
rev = 0

while num > 0:
    digit = num % 10
    rev = rev * 10 + digit
    num //= 10

if original == rev:
    print("The number is a palindrome")
else:
    print("The number is not a palindrome")


Enter a number: 121
The number is a palindrome


In [ ]:
s = input("Enter a string: ")

left = 0
right = len(s) - 1

is_palindrome = True

while left < right:
    if s[left] != s[right]:
        is_palindrome = False
        break
    left += 1
    right -= 1

if is_palindrome:
    print("The string is a palindrome")
else:
    print("The string is not a palindrome")


Enter a string: abba
The string is a palindrome


Create a menu:

1. Say Hello
2. Say Bye
3. Exit

Run the code until the user press Exit.

In [ ]:
n = 0
print("\nMenu:")
print("1. Say Hello")
print("2. Say Bye")
print("3. Exit")
while (n != 3):
  n = int(input("Enter your option (1/2/3): "))
  if n == 1:
    print("Hello")
  elif n== 2:
    print("Bye")
  elif n ==3:
    print("Exit")
  else:
    print("Wrong Choice")





Menu:
1. Say Hello
2. Say Bye
3. Exit
Enter your option (1/2/3): 1
Hello
Enter your option (1/2/3): 2
Bye
Enter your option (1/2/3): 4
Wrong Choice
Enter your option (1/2/3): 3
Exit


For Loop

A for loop is used for iterating over a sequence (that is either a list, a tuple, a dictionary, a set, or a string).

This is less like the for keyword in other programming languages, and works more like an iterator method as found in other object-orientated programming languages.

With the for loop we can execute a set of statements, once for each item in a list, tuple, set etc.

The for loop is used when you know in advance how many times you want to execute the block of code.
It iterates over a sequence (e.g., a list, tuple, string, or range) and executes the block of code for each item in the sequence.
The loop variable (variable) takes the value of each item in the sequence during each iteration.

The while loop is used when you don't know in advance how many times you want to execute the block of code. It continues to execute as long as the specified condition is true.
It's important to make sure that the condition eventually becomes false; otherwise, the loop will run indefinitely, resulting in an infinite loop.

In [ ]:
fruits = ["apple", "banana", "cherry"]    #syntax
for x in fruits:
  print(x)

apple
banana
cherry


Iterating over sets and accessing set elements

In [ ]:
thisset = {"apple", "banana", "cherry"}

for x in thisset:
  print(x)

cherry
banana
apple


In [ ]:
thisset = {"apple", "banana", "cherry"} #Check if "banana" is present in the set:

print("banana" in thisset)

True


In [ ]:
thisset = {"apple", "banana", "cherry"}   #Check if "banana" is NOT present in the set:

print("banana" not in thisset)

Add Items to set

In [ ]:
thisset = {"apple", "banana", "cherry"}

thisset.add("orange")

print(thisset)


{'cherry', 'banana', 'orange', 'apple'}


In [ ]:
thisset = {"apple", "banana", "cherry"}
tropical = {"pineapple", "mango", "papaya"}

thisset.update(tropical)

print(thisset)


{'mango', 'cherry', 'banana', 'pineapple', 'papaya', 'apple'}


In [ ]:
thisset = {"apple", "banana", "cherry"}
mylist = ["kiwi", "orange"]

thisset.update(mylist)

print(thisset)

{'cherry', 'banana', 'kiwi', 'orange', 'apple'}


Remove Item
To remove an item in a set, use the remove(), or the discard() method.

In [ ]:
thisset = {"apple", "banana", "cherry"}

thisset.remove("banana")

print(thisset)


{'cherry', 'apple'}


In [ ]:
thisset = {"apple", "banana", "cherry"}

thisset.discard("banana")

print(thisset)

{'cherry', 'apple'}


You can also use the pop() method to remove an item, but this method will remove a random item, so you cannot be sure what item that gets removed.

The return value of the pop() method is the removed item.

In [ ]:
thisset = {"apple", "banana", "cherry"}

x = thisset.pop()

print(x)

print(thisset)

cherry
{'banana', 'apple'}


In [ ]:
thisset = {"apple", "banana", "cherry"}   #The clear() method empties the set

thisset.clear()

print(thisset)

set()


In [ ]:
thisset = {"apple", "banana", "cherry"}
del thisset
print(thisset)


NameError: name 'thisset' is not defined

Join Sets
There are several ways to join two or more sets in Python.

The union() and update() methods joins all items from both sets.

The intersection() method keeps ONLY the duplicates.

The difference() method keeps the items from the first set that are not in the other set(s).

The symmetric_difference() method keeps all items EXCEPT the duplicates.

Union
The union() method returns a new set with all items from both sets

In [ ]:
set1 = {"a", "b", "c"}
set2 = {1, 2, 3}

set3 = set1.union(set2)
print(set3)

In [ ]:
set1 = {"a", "b", "c"}
set2 = {1, 2, 3}

set3 = set1 | set2  #You can use the | operator instead of the union() method, and you will get the same result
print(set3)

Join Multiple Sets
All the joining methods and operators can be used to join multiple sets.

When using a method, just add more sets in the parentheses, separated by commas:

In [ ]:
set1 = {"a", "b", "c"}
set2 = {1, 2, 3}
set3 = {"John", "Elena"}
set4 = {"apple", "bananas", "cherry"}

myset = set1.union(set2, set3, set4)
print(myset)

{1, 2, 3, 'cherry', 'a', 'c', 'b', 'Elena', 'bananas', 'John', 'apple'}


In [ ]:
set1 = {"a", "b", "c"}
set2 = {1, 2, 3}
set3 = {"John", "Elena"}
set4 = {"apple", "bananas", "cherry"}

myset = set1 | set2 | set3 |set4
print(myset)

{1, 2, 3, 'cherry', 'a', 'c', 'b', 'Elena', 'bananas', 'John', 'apple'}


oin a Set and a Tuple
The union() method allows you to join a set with other data types, like lists or tuples.

The result will be a set.

In [ ]:
x = {"a", "b", "c"}
y = (1, 2, 3)

z = x.union(y)
print(z)
print(type(z))

{1, 2, 3, 'a', 'c', 'b'}


In [ ]:
x = {"a", "b", "c"}
y = [1, 2, 3]

z = x.union(y)
print(z)
print(type(z))

{1, 2, 3, 'a', 'c', 'b'}
<class 'set'>


Update
The update() method inserts all items from one set into another.

The update() changes the original set, and does not return a new set.

In [ ]:
set1 = {"a", "b" , "c"}
set2 = {1, 2, 3}

set1.update(set2)
print(set1)

{'b', 1, 2, 3, 'a', 'c'}


Python frozenset
frozenset is an immutable version of a set.

Like sets, it contains unique, unordered, unchangeable elements.

Unlike sets, elements cannot be added or removed from a frozenset.

In [ ]:
x = frozenset({"apple", "banana", "cherry"})
print(x)
print(type(x))

frozenset({'cherry', 'banana', 'apple'})
<class 'frozenset'>


Assignment Time!!

Create an empty set called data_set.

Insert the elements 10, 20, 30, and 40 into the set.

Display the set.

Delete the element 20 from the set.

Display the set after deletion.

Remove one element from the set using pop() and display:

The removed element

The set after popping

Create another set extra_set containing 50 and 60.

Update data_set by adding all elements of extra_set.

Display the updated set.

Join data_set with another set {30, 70, 80} and store the result in joined_set.

Display the joined set.

Clear all elements from data_set.

Display the final empty set.

In [ ]:
# Single program demonstrating multiple set operations

# Step 1: Create an empty set
data_set = set()

# Step 2: Insert elements
data_set.add(10)
data_set.add(20)
data_set.add(30)
data_set.add(40)

print("Initial set:", data_set)

# Step 3: Delete an element
data_set.remove(20)
print("After deletion:", data_set)

# Step 4: Pop an element
removed_element = data_set.pop()
print("Popped element:", removed_element)
print("Set after pop:", data_set)

# Step 5: Update the set
extra_set = {50, 60}
data_set.update(extra_set)
print("After update:", data_set)

# Step 6: Join two sets (Union)
another_set = {30, 70, 80}
joined_set = data_set.union(another_set)
print("Joined set:", joined_set)

# Step 7: Clear the set
data_set.clear()
print("After clear:", data_set)


Initial set: {40, 10, 20, 30}
After deletion: {40, 10, 30}
Popped element: 40
Set after pop: {10, 30}
After update: {50, 10, 60, 30}
Joined set: {80, 50, 70, 10, 60, 30}
After clear: set()


For loop Continues

In [ ]:
for x in "banana":          #looping through a string
  print(x)

b
a
n
a
n
a


The break Statement
With the break statement we can stop the loop before it has looped through all the items:

In [ ]:
fruits = {"apple", "banana", "cherry"}
for x in fruits:
  print(x)
  if x == "banana":
    break

cherry
banana


In [ ]:
fruits = {"apple", "banana", "cherry"}
for x in fruits:
  if x == "banana":
    break       #Exit the loop when x is "banana", but this time the break comes before the print:
  print(x)

cherry


Range

Python range
The built-in range() function returns an immutable sequence of numbers, commonly used for looping a specific number of times.

This set of numbers has its own data type called range.

Note: Immutable means that it cannot be modified after it is created.
Creating ranges
The range() function can be called with 1, 2, or 3 arguments, using this syntax:

range(start, stop, step)

Call range() With One Argument
If the range function is called with only one argument, the argument represents the stop value.


The start argument is optional, and if not provided, it defaults to 0.

range(10) returns a sequence of each number from 0 to 9. (The start argument, 0 is inclusive, and the stop argument, 10 is exclusive).

Call range() With Two Arguments
If the range function is called with two arguments, the first argument represents the start value, and the second argument represents the stop value.

range(3, 10) returns a sequence of each number from 3 to 9

Call range() With Three Arguments
If the range function is called with three arguments, the third argument represents the step value.

The step value means the difference between each number in the sequence. It is optional, and if not provided, it defaults to 1.

range(3, 10, 2) returns a sequence of each number from 3 to 9, with a step of 2

Example of Range using for loop

Write a Python program to:

Print numbers from 1 to 10 using a for loop.

In [ ]:
for i in range(1, 11):
    print(i)


1
2
3
4
5
6
7
8
9
10


Write a Python program to:

Take a number N from the user.

Calculate and print the sum of first N natural numbers using a for loop.

In [ ]:
num = int(input("Enter the number till which you want the summation: "))
sum = 0
for x in range (1, num+1):
  sum = sum+x
print(sum)


Enter the number till which you want the summation: 4
1
3
6
10


Write a Python program to:

Take a number from the user.

Print the multiplication table of that number from 1 to 10 using a for loop.

In [ ]:
num = int(input("Enter a number: "))

for i in range(1, 11):
    print(num, "x", i, "=", num * i)


Enter a number: 2
2 x 1 = 2
2 x 2 = 4
2 x 3 = 6
2 x 4 = 8
2 x 5 = 10
2 x 6 = 12
2 x 7 = 14
2 x 8 = 16
2 x 9 = 18
2 x 10 = 20


Write a Python program to:

Take a list of numbers from the user. (hint: use while loop)

Use a for loop to count:

Total even numbers

Total odd numbers

Display both counts.

In [ ]:
x = int(input("Enter the number of elements in the list: "))   #creating a list using user input

list1 = []   #initialize the empty list
i = 0  #initializing while loop variable

while i <= x:
    num = int(input("Enter a number: "))
    list1.append(num)
    i = i+1

print(list1)


Enter the number of elements in the list: 6
Enter a number: 11
Enter a number: 12
Enter a number: 4
Enter a number: 6
Enter a number: 5
Enter a number: 14
[11, 12, 4, 6, 5, 14]


In [ ]:
even = 0 #counter of even number
odd = 0   #counter of odd number
for x in list1:  #iterating through the list
  if x%2 == 0:
    even = even+1
  else:
    odd = odd+1
print("even= ", even, "odd= ",odd)

even=  4 odd=  2


Write a Python program to:

Take a set of numbers from the user using a while loop.

Use a for loop to count:

Total positive numbers

Total negative numbers

Total zeros

Display all three counts.

In [ ]:
x = int(input("Enter the number of elements in the list: "))   #creating a list using user input

set1 = set()
i = 0

while i < x:
    num = int(input("Enter a number: "))
    set1.add(num)
    i = i+1

print(set1)


Enter the number of elements in the list: 6
Enter a number: 0
Enter a number: 1
Enter a number: -1
Enter a number: -2
Enter a number: 4
Enter a number: -4
{0, 1, 4, -2, -4, -1}


In [ ]:
pos=0
neg=0
zer = 0
for i in set1:
  if i>0:
    pos = pos+1
  elif i<0:
    neg = neg+1
  else:
    zer = zer+1
print("postives: ", pos, "Negatives= ", neg, "Zero= ", zer)

postives:  2 Negatives=  3 Zero=  1


Write a Python program to:

Take a list of numbers from the user.

Use a for loop to find:

The largest number in the list

The smallest number in the list

Display both values.

In [ ]:
x = int(input("Enter the number of elements in the list: "))   #creating a list using user input

list1 = []
i = 0

while i < x:
    num = int(input("Enter a number: "))
    list1.append(num)
    i = i+1

print(list1)


Enter the number of elements in the list: 3
Enter a number: 4
Enter a number: 22
Enter a number: 1
[4, 22, 1]


In [ ]:
smallest = list1[0]
largest = list1[0]

for x in list1:
    if x < smallest:
        smallest = x
    if x > largest:
        largest = x

print("smallest:", smallest)
print("largest:", largest)


smallest: 1
largest: 22


Write a Python program to:

Take an integer number from the user.

Use a loop to reverse the number.

Display the reversed number.

check whether it is a palindrome or not.

In [ ]:
num = int(input("Enter a number: "))
num = str(num)
reverse = ""

for digit in num:
    reverse = digit + reverse
print(reverse)

if num == reverse:
    print("Palindrome number")
else:
    print("Not a palindrome number")


Enter a number: 124
421
Not a palindrome number


Write a Python program to:

Use a for loop to print the square of numbers from 1 to 10.

In [ ]:
for x in range(1, 11):
  print ("The Square of x is: ", x*x)

The Square of x is:  1
The Square of x is:  4
The Square of x is:  9
The Square of x is:  16
The Square of x is:  25
The Square of x is:  36
The Square of x is:  49
The Square of x is:  64
The Square of x is:  81
The Square of x is:  100


Write a Python program to:

Take two numbers from the user (start and end).

Use a for loop to calculate the sum of all even numbers between them.

In [ ]:
num1 = int(input("Enter a number: "))
num2 = int(input("Enter a number: "))
sum = 0
for i in range(num1, num2+1):
  if i%2 == 0:
    sum = sum+i
print("Sum of all even numbers: ",sum)

Enter a number: 1
Enter a number: 10
Sum of all even numbers:  30


Write a Python program to:

Take a string from the user.

Use a for loop to count the number of vowels in the string.

In [ ]:
string1 = str(input("Enter the string: "))
count = 0
for i in string1:
  if i == 'a':
    count = count+1
  elif i == 'e':
    count = count+1
  elif i == 'i':
    count = count+1
  elif i == 'o':
    count = count+1
  elif i == 'u':
    count = count+1
print(count)

Enter the string: aaaiiiiuuuoooeeeuuffgghhaaiiuuooeehhuuiiddooeekkeelloommee
42


Write a Python program to:

Take a number from the user.

Use a for loop to calculate the factorial of the number.

📌 Example: 5! = 5 × 4 × 3 × 2 × 1 = 120

In [ ]:
num = int(input("Enter the number: "))
fact = 1
if num==0:
  print(fact)
for i in range(1, num+1):
  fact = fact*i
print(fact)

Enter the number: 5
120


Nested for loops

In [ ]:
for i in range(1, 10):
  for j in range(1, 10):
    print(i+j)

2
3
4
5
6
7
8
9
10
3
4
5
6
7
8
9
10
11
4
5
6
7
8
9
10
11
12
5
6
7
8
9
10
11
12
13
6
7
8
9
10
11
12
13
14
7
8
9
10
11
12
13
14
15
8
9
10
11
12
13
14
15
16
9
10
11
12
13
14
15
16
17
10
11
12
13
14
15
16
17
18


In [ ]:
fruits =["apple", "banana", "watermelon"]
cars=["volvo", "benz", "tata", "skoda"]
for i in fruits:
  for j in cars:
    # print(i)
    # print(j)
    print(i+" "+j)

apple volvo
apple benz
apple tata
apple skoda
banana volvo
banana benz
banana tata
banana skoda
watermelon volvo
watermelon benz
watermelon tata
watermelon skoda



Write a Python program using for loops to print:

1

1 2

1 2 3

1 2 3 4

1 2 3 4 5

In [ ]:
for i in range(1, 6):
    for j in range(1, i + 1):
        print(j, end=" ")
    print()
#Outer loop (i) → controls the number of rows

# Inner loop (j) → prints numbers from 1 to i

# end=" " → prints numbers on the same line

# print() → moves to the next line after each row

1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


Write a Python program using nested for loops to print:

'*' in a similar way

In [ ]:
for i in range(1, 6):
    for j in range(1, i + 1):
        print("*", end=" ")
    print()


* 
* * 
* * * 
* * * * 
* * * * * 


Write a Python program using nested for loops to print:

1

2 2

3 3 3

4 4 4 4

5 5 5 5 5

In [ ]:
for i in range(1, 6):
    for j in range(1, i + 1):
        print(i, end=" ")
    print()

# Outer loop (i) → controls the number printed

# Inner loop (j) → controls how many times it is printed

# end=" " → prints on the same line

# print() → moves to the next line


1 
2 2 
3 3 3 
4 4 4 4 
5 5 5 5 5 


Write a Python program to print the multiplication tables from 1 to 5 in a grid format using nested for loops.

📌 Sample Output (partial):

1  2  3  4  5

2  4  6  8  10

3  6  9  12 15

...............

In [ ]:
for i in range(1, 6):
    for j in range(1, 6):
        print(i * j, end="\t")
    print()
# Outer loop (i) → rows (1 to 5)

# Inner loop (j) → columns (1 to 5)

# i * j → multiplication result

# \t → keeps output aligned in grid form

# print() → moves to next row

1	2	3	4	5	
2	4	6	8	10	
3	6	9	12	15	
4	8	12	16	20	
5	10	15	20	25	


Python Dates
A date in Python is not a data type of its own, but we can import a module named datetime to work with dates as date objects.

In [ ]:
import datetime

x = datetime.datetime.now()
print(x)

Date Output
When we execute the code from the example above the result will be:

2026-02-11 00:34:36.314985
The date contains year, month, day, hour, minute, second, and microsecond.

The datetime module has many methods to return information about the date object.

In [ ]:
import datetime

x = datetime.datetime.now()

print(x.year)
print(x.strftime("%A"))

2026
Tuesday


The strftime() Method
The datetime object has a method for formatting date objects into readable strings.

The method is called strftime(), and takes one parameter, format, to specify the format of the returned string:

In [ ]:
import datetime

x = datetime.datetime(2018, 6, 1)

print(x.strftime("%B"))

June


To create a date, we can use the datetime() class (constructor) of the datetime module.

The datetime() class requires three parameters to create a date: year, month, day.

In [ ]:
import datetime

x = datetime.datetime(2020, 5, 17)

print(x)

2020-05-17 00:00:00
